### PR TITLE
server support for system, prefix, and suffix prompts with special to…

### PR DESCRIPTION
**Pull request by ghost**
The example/server cannot support instruction tuned models which require special tokens to function correctly such as Mistrallite.  To solve this I added support to the prompt tokenizer in the server to support adding a system, input_prefix, and input_suffix to text prompts with special token support.  I also added a "tokenizes" POST to support tokenizing special tokens being used by new models with chatML such as <|im_start|> <|im_end|> <|prompter|> etc. to help verify the special tokens are correct.  The tokenizer logic assumes an input text array prompt is a chat history toggling between user input lines and model output lines and adds the approprate prefix/suffix to the prompt lines to form a valid character sequence with the needed special token prompt delimiters for the model.  This functionality was extensive tested across 7 instruction tuned Mistral 7B chat models and the falcon 180B chat model and found to work well with all models giviing good outputs on chat histories as opposed to either not working at all or giving goofy outputs without the correct prompt delimiters needed by the model.  The functionality was tested extensively over the weekend but today I found major changes were made to the server code with the slotting etc. so I rewrote quickly against the new slotted code but it still seems to work OK.

### CODE REVIEW

File: examples/server/server.cpp [hunk 1]
```diff
@@ -49,10 +49,10 @@ static bool server_verbose = false;
...
     do                                                                   \
     {                                                                    \
-        if (server_verbose)                                              \
-        {                                                                \
-            server_log("VERBOSE", __func__, __LINE__, MSG, __VA_ARGS__); \
-        }                                                                \
+	if (server_verbose)                                              \
+	{                                                                \
+	    server_log("VERBOSE", __func__, __LINE__, MSG, __VA_ARGS__); \
+	}                                                                \
```
Comment (cebtenzzre) [inline examples/server/server.cpp:55]: This project uses spaces for indentation. Do not replace them with tabs.

**Comment by shibe2** [comment]
Different models use different formats, some of them don't fit into system/prefix/suffix model. Prompt formatting is currently done on client, and server already supports prompts with special tokens.

I noticed that you use `special` option of tokenizer when processing system, input_prefix and input_suffix. As I understand, it is meant for examples like main, where there is no other way to supply special tokens. This method has a shortcoming in that it cannot distinguish between special token like im_start and literal text "<|im_start|>". I'd prefer not having this in server API to discourage users from handling special tokens in that way. server has a different, better way of supplying special tokens by their numeric identifiers, as in this example: https://github.com/ggerganov/llama.cpp/discussions/3599#discussioncomment-7284649

**Comment by alt8git** [comment]
Before this patch I could not figure out any way to get the system prompt and prefix/suffix added correctly to the prompts with special tokens through the server interface.  It is necessary to only use special tokens on these three, special tokens should not be used when tokenizing the user prompt itself.  Every mistral chat model I tried requires these prompts be added to the user prompt.  Main is the only interface which would run mistrallite correctly because it does tokenize prefix/suffix with special tokens and concatenates them exactly as I do in the patched server tokenizer.  This new function adds the same capability main has in server so every mistral model I tried now works right through server.  All the mistral models and falcon fit into the system/prefix/suffix model and were tested to function right with the patch.  The patch also enables a new feature in that generic chat histories can easily shared between different models so the models can be short-term primed on a generic chat history correctly while the server tokenizer automatically handles correctly adding the required prefix/suffixes to the chat.  The patch also does not force the user to use it, it is passive in that if the input prefix, input suffix, and system are not specified the interface works just like it used to.

I would make one change to my patch in that I would pass a new parameter "special" to the tokenize function instead of creating a separate tokenizes POST, this is cleaner.  This option is needed to tokenize system prompt with special tokens and get the correct number of tokens for NKEEP, and also helps with debugging.

File: examples/server/server.cpp (hunk 1 above)
Comment (alt8git) [reply examples/server/server.cpp]: fixed.  my editor defaults to tab compression on write.

**Review by goerch — CHANGES_REQUESTED** [review]
If you are touching tokenization I'd like to see a lot of tests for it.

Edit: I see. it is server side tokenization, don't mind.

**Review by goerch — COMMENTED** [review]
Sorry for the noise.

**Comment by alt8git** [comment]
> Sorry for the noise.

As mentioned in PR this patch is to enable server to support special token generation needed by some newer instruction tuned models (mistral, mistrallite) for system, prompt prefix and prompt suffix using either chatML or custom special tokens such as <|prompter|> and <|assistant>| used by mistrallite (the model (s) will not work without use of these special tokens).  The user should not be required to decipher these tokens externally to numbers and pass them in manually as ints, it is cleaner to pass the ascii string and let the tokenizer handle translating the strings to tokens internally to the server.

**Pull request closed without merging**